### DISCUSSION TITLE
Controlling multiple desks/legs

### DISCUSSION THREAD

**Discussion by Nikspace2** [gcormier/megadesk#55] [Ideas]
Hi guys, I am happy owner of 2 Bekant desks wich i have extended them and combined them in a L shape desk.

Until now I have been operating them by pressing bothbuttons at the shame time. If they get misaligned is just lower them to the bottom where two IKEA ALEX cabinets on the ends and a desk leg on the corner are holding the Bekant legs on the air.

![IMG_20190409_194916](https://user-images.githubusercontent.com/7294372/116013501-86656400-a5fe-11eb-82e0-512ba13bae2e.jpg)

I have tried using a arduino to sync this action. But I only manage with just closing circuit of the buttons. I thought using couple ultrasonic distance sensors. But then I found a tut. on youtube that someone was reading the LIN bus to get the height of the desk. I try doing that but I didn't know how to read from 2 different units with a single arduino. 

Then I found megadesk, but I didn't have the time to look into them. But now I want to look into that again. And I was thinking if I could ask a custom version of megadesk. 

That would be awesome, then I could expand on that version and print an custom enclosure with memory buttons and why not even use my elgato stream deck.
Do be honest I am not interested of using the original enclosure or buttons at all.

But first thing first, how we could read from two units and then send the same command on both.

If we could read both units, then we could also see if they are misaligned and even balance them automatically.

I just throwing ideas here. I am a marine engineer and hobbyist C# programmer. So my knowledge in electronics is a bit limited. I know just the basics with an arduino.

If gcormier willing to make a custom version of something like that, of course I am willing of paying any extras RnD and work he will do for something like that.

Thank you for reading my idea.

**Comment by gcormier** [comment]
Wow man. This, truly, is a MEGADESK.

It is definitely possible. The hardware would probably be pretty straightforward, the software a bit more involved.

Hardware wise, a microcontroller with 2 serial ports, and then 2 LIN bus transceivers. You would need to have seperate LIN bus, as the motor's will probably share the same ID for left and right (eg 2 left motors, 2 right motors)... it would be interesting to see what happens if you tried with 1 LIN bus but I can't imagine it being very stable. If one motor drifts position at all it could cause issues.

It might also be possible with 2 current megadesks, and the 2 are running different firmwares. One becomes primary that controls everything - the 2nd board relays the information over serial bus to the primary and decides what to do. The buttons are only connected to the primary unit.

Someone has already merged the code for serial communications. One of the challenges might be the limited amount of memory available on the ATTiny841.

Let me think about it for a bit see what ideas I come up with :)

**Reply by Nikspace2** [reply]
Take your time, as I said I work in ships, so until August i will not be back home.

**Reply by tagno25** [reply]
One possible solution is to use middle microcontroller to manage a cluster.  It would either have multiple serial ports and/or use soft serial.  The managing controller could be setup to convert the serial commands bi-directional, so any controller could change the heights for all of the controllers or just have a single master.

If setup for bi-directional, you would need to somehow determine which controller is currently changing the height.

Depending on how the managing controller is setup, you could either have the same memory heights on every controller , or even have the controllers save different memory heights (Ex. controller 1 memory 2 sets all to 800mm, but controller 2 memory 2 sets all to 1000mm).  As long as custom min/max is not enabled, then the height change is updated throughout the memory moves (may or may not be necessary). 

Unfortunately I do not have two desks that I am able to connect to each other and make/test any multi-desk code.

**Reply by gcormier** [reply]
Without any custom circuit design, I agree - a master controller would be best. Any sort of Arduino-style board with two serial ports would be preferable, and it can talk to the 2 megadesk boards. Each would just listen to commands from the master node and would otherwise not really know they are being orchestrated :)

Surely there's many ways to tackle the issue as well with hardware, but that will probably be your quickest path as it will only rely on software/programming which you will be a bit more familiar with then hardware design.

**Reply by Nikspace2** [reply]
Ok then i will order 2 megadesks. 
And until i receive them, i will start looking into the serial branch.
To see what info i can share between the megadesk and the microprocessor.

**Reply by Mr-HaXx** [reply]
Any update?

**Comment by hhoef** [comment]
Is there any news on controlling two desks simultaneously? I have the same problem and would love to order 2 megadesk boards to solve it. Also I'm willing to pay extra for the R&D.

**Comment by gcormier** [comment]
Biggest challenge here is having 2 desks to play with :) Otherwise the implementation would be straightforward with the serial protocol. If anyone has spare (working) legs they want to ship, that could work as well.

**Comment by kobayashimaru1701** [comment]
Very much interested in this project. I can't find anything online about connecting multiple desks. I was planning on implementing a rudimentary hardware solution to essentially "press" both sets of buttons but would gladly pay for a single controller that can natively handle it.

**Comment by Mr-HaXx** [comment]
Seems simple soldering of the original boards to controll 4 legs didn't work. That's what I tried first without searching online. It didn't work. Has anyone managed to get anything to work? Also I have 0 experience in Arduino and setting up controls.

**Reply by gcormier** [reply]
It depends on how the hardware works in the IKEA desk.

First, if you can check the stickers on each leg to see if they have any differences, this would imply that when manufactured, there is a left leg and a right leg. This would then strongly indicate that the ability to control 4 legs must be done by software and 2 different busses.

If they are identical, it is possible there is internal stuff on the PCB, or something in the firmware - but the factory/assembly plant needs workers to know which 2 legs to put in each box before shipping.

So I would suggest disassembly as a 1st step to check if they have anything unique. part number, barcode, notes, stickers, etc.

**Reply by Mr-HaXx** [reply]
From what I see is they each have a different serial number printed on them. And each internal motor has a serial number. All 4 legs have a completely different serial number.
I attached a picture of the motor inside of the legs
![17036142331723267413654060908214](https://github.com/gcormier/megadesk/assets/29613887/f4ebcec0-5640-4251-bc47-b7fdb22fde9c)

**Reply by gcormier** [reply]
Do they have different part numbers?

**Reply by Mr-HaXx** [reply]
Only thing different are the numbers shown on the picture. Partnumbers are all ikeapartnumber: 502.552.54 which is the code for these legs on the website

**Reply by Mr-HaXx** [reply]
And this is the only sticker visible on the outside of the legs
![17036150562451265936717805516931](https://github.com/gcormier/megadesk/assets/29613887/8f7f2da6-9014-40a1-82ef-eabe29d5970a)

**Reply by gcormier** [reply]
Again, the premise is the legs need to be put in boxes by humans. If they are different, then they must have some sort of indications - one can't rely on "not mixing up the legs".

So if they are indeed identical (not requiring disassembly of the leg itself), then it's likely we can try to control these from 1 megadesk. They might have some sort of random PID assignment logic. 1-Wire has the ability to have devices create PID's on the fly. LIN doesn't specify this, but it could be a custom addon.

You could place a logic analyzer (eg https://www.amazon.ca/Comidox-Analyzer-Device-Channel-Arduino/dp/B07KW445DJ ) on the LIN data bus to see what is being sent when all 4 motors are attached.

**Reply by Mr-HaXx** [reply]
I will order that device. And get back in here for instructions on how to use

**Reply by Mr-HaXx** [reply]
Oke so I bought the device and it has arrived. I have 0 knowledge on how to use it. Can you give me a simple guide or help me through it ao I can share the info here?

**Reply by gcormier** [reply]
Google will bring you a plethora of results in your favorite media of choice - text or video. eg https://learn.sparkfun.com/tutorials/using-the-usb-logic-analyzer-with-sigrok-pulseview or https://www.youtube.com/watch?v=rR5cEFRO9_s or many others.

You can use the Saleae software or Pulseview/Sigrok and save/post the captures here. You'll need to tap the GND and LIN connections from the cable - be careful not to short anything out!

**Comment by kobayashimaru1701** [comment]
For what it's worth, I ended up going with a low tech solution. I rigged up
a relay system and wired it to 2 controllers in order to effectively
control 4 legs simultaneously. It isn't perfect but it works for now.

Fyi I'm pretty sure all of the legs are identical. I've mixed and matched
legs to controllers (2 at a time) and they always work as expected.

**Reply by Mr-HaXx** [reply]
Yeah the legs are identical, but I assume the controller can only read 2 lege, and if its anything other then 2 legs it won't work. I guess it's a failsafe. But I'm here to still make it work. I just need some basic knowledge on how boards work and how to read them

**Comment by kobayashimaru1701** [comment]
Just a fyi I burned out one set of legs due to a goof on my part so I got another pair. I noticed the circuit board is marked as a different revision: "Rev. 10.9 Soft 3.8 03/17"
My previous sets were simply marked as "Revision 2" or "Revision 3".
The general design appears to be similar but the IC layout is different between them. Probably immaterial to this discussion but thought I'd share the variants I've seen.

**Reply by cerahmed** [reply]
Hey thanks for sharing your experience. Can you please provide more details on how you went about wiring the relay system? I'm about to attempt the same.

**Comment by Mr-HaXx** [comment]
I have tried a lot, the problem is that the chip in the Ikea table is programmed to only work with 2 legs, it checks if it receives a pulse from both legs, and checks if the voltage is correct. So when I soldered extra legs, it stopped working. So I gave up, used 1 pair of legs with 1 controller, and just put 2 desks of 1,50m on top, and used reinforced metal underframe I welded. So the tables lifts the 3m table with no problems and supports all the weight on top.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by kobayashimaru1701** [comment]
I wired one SPDT relay to control the UP signal for both legs and a
separate relay to control the DOWN signal.It's pretty rudimentary and it
occasionally hiccups but it works.

The main issue is when one pair of legs doesn't respond for whatever
reason. You have to watch for this each time because you dont want the
height difference to get too extreme.
Also the legs arent perfectly in sync. Ive noticed that there are several
different revisions of the controller board so this might have to do with
it. However its close enough that it isnt a problem for me.

**Reply by cerahmed** [reply]
Do you have photos of how you managed to wire the relay to the up/down buttons? Did you solder intot the small pins in the PCB?
Also, what hiccups did you face? I'm thinking of having a single table top extending the 4 legs, from your experience would I have issues?

**Edit**: received my Bekant legs, and have done some progress (the Arudino in the video has nothing to do with the controllers, it just sits their next to the bread board 😁).


https://github.com/user-attachments/assets/2bc96bee-36ab-430c-b396-cbb98a56d6a5



I have connected each of the three wires (UP, DOWN, GND) from both controllers together, and tried pulling the UP and DOWN wires to GND; it worked as expected. However I can see (and hear the clicks of the motors' relays) that the pair of legs on the left is almost always half a second faster than the one on the right - in both directions. Switching controllers between the legs still yielded the same result, so it has to do with the motors themselves.

I don't have the custom desk top connected yet, so I don't know if that would be an issue. Will do some tests when I get the top ready.

**Reply by kobayashimaru1701** [reply]
Sorry I don't check e-mail often and just saw this now.

Hmmm it's been a while so my memory is a little hazy but I recall running into issues connecting the controllers directly to each other like that. So instead I used a double-pole switch and then later replaced that with relays triggered by a wireless remote.

As for the wiring, I went super ghetto and just stripped the ends off some thing gauge wire and shoved them into the existing ribbon connector in the original controller. I then hot glued it in place. It was supposed to be "temporary" but we all know how that goes lol.

I couldn't quite figure out how to get the two pairs to be in perfect sync with each other. I tried swapping different combinations of internal hardware thinking there might be some difference. I even tried altering the length of wire to make it as even as possible but in the end nothing worked. I figure it must be a calibration thing which would be difficult to sync since they don't talk to each other.
The main annoyance I've faced is when one pair of legs stops responding for whatever reason. A power cycle usually fixes the problem but hopefully you stopped before the whole thing is too far off kilter.
